rfcs/text/0006_management_section_service.md

```diff
@@ -0,0 +1,284 @@
+2. Expose a registry for other plugins to add their own registry sections to the UI and add nested links to them in the sidebar.
+
+The purpose of this RFC is to consider item 2 above -- the service for registering sections to the nav & loading them up.
+
+# Motivation
+
+## Why now?
+The main driver for considering this now is that the Management API moving to the new platform is going to block other teams from completing migration, so we need to have an answer to what the new platform version of the API looks like as soon as possible in `7.x`.
+
+## Why not just keep the current API and redesign later?
+The answer to that has to do with the items that are currently used in the management implementation which must be removed in order to migrate to NP: the framework currently registers a `uiExport`, and relies on `IndexedArray`, `uiRegistry`, and `ui/routes`.
+
+This means that we will basically need to rebuild the service anyway in order to migrate to the new platform. So if we are going to invest that time, we might as well invest it in building the API the way we want it to be longer term, rather than creating more work for ourselves later.
+
+## Technical goals
+- Remove another usage of `IndexedArray` & `uiRegistry` (required for migration)
+- Remove dependency on `ui/routes` (required for migration)
+- Remove management section `uiExport` (required for migration)
+- Simple API that is designed in keeping with new platform principles
+  - This includes being rendering-framework-agnostic... You should be able to build your management section UI however you'd like
+- Clear separation of app/UI code and service code, even if both live within the same plugin
+- Flexibility to potentially support alternate layouts in the future (see mockups in [reference section](#reference) below)
+
+# Basic example
+This API is influenced heavily by the [application service mounting RFC](https://github.com/elastic/kibana/blob/master/rfcs/text/0004_application_service_mounting.md). The intent is to make the experience consistent with that service; the Management section is basically one big app with a bunch of registered "subapps".
+
+```ts
+// my_plugin/public/plugin.ts
+
+export class MyPlugin {
+  setup(core, { management }) {
+    management.sections.register({
+      id: 'my-section',
+      title: 'My Main Section', // display name
+      description: 'hello', // not used in current UI, but possibly in future
+      order: 10,
+      euiIconType: 'iconName',
+    });
+    management.sections.registerLink('my-section', {
+      id: 'my-link',
+      title: 'My Link', // display name
+      description: 'hello', // not used in current UI, but possibly in future
+      order: 20,
+      async mount(context, params) {
+        const { renderApp } = await import('./my-section');
+        return renderApp(context, params);
```

Comment:
As I understand the Management App will be written in React and all sections will be written in React, too. Maybe this should receive a React component, instead of unmounting React and mounting it back.

Reply:
Anyone who does not want to use React can easily use whatever technology they want:

```jsx
const AppSection: React.FC = () => {
  const ref = useRef();
  useLayoutEffect(() => {
    if (ref.current) renderWhateverOtherTechnology(ref.current);
  }, [ref.current]);
  return <div ref={ref} />;
};
```

Reply:
> Counterpoint: Any plugin that wants to register a section without using React would have to bundle React in their plugin just do these 5 lines.

React will be provided globally to everyone, you cannot have more than one version of React in a single React tree.

Reply:
I suggest we look at this topic from technical perspective&mdash;instead of blindly copying what Core does for application mounting and "I like"&mdash;let's look at what we gain and lose by re-mounting React.

- This React re-mounting adds unnecessary code: (1) un-mount React; (2) mount it back; (3) add tests for all of that&mdash;which needs to be written and then maintained. Then you need to (4) re-attach all your Error Boundaries; (5) re-attach all your React contexts (which I don't think is possible, in a non-trivial cases); (6) all your React Suspense handlers will be probably broken (I'm not aware if it is possible to re-connect them).
- Re-mounting React *does not make us less dependent on React*: in both cases the Management app is written in React and all the sections are written in React. The only difference is, if we do React re-mounting, we need to **add extra code**, which makes the API harder to use and maintain (and is slower and breaks React Context and breaks React Error Boundaries and breaks React Suspense).
  - Management app and all its sections are written in React, by adding a few % of code that re-mounts React does not make us any less React dependent and those few % of code are not needed in the first place.
- Un-mounting React and then mounting it back *breaks [React Context](https://reactjs.org/docs/context.html)*. (cc @ppisljar @stacey-gammon)
  - For example, if we ever had dynamic theming in EUI, where theme is stored in React context, in Kibana we would not be able to use it, at least in a sensible way, because of this React context breaking. (cc @elastic/kibana-design)
  - Similar goes for i18n&mdash;i18n context has to be re-attached if you want to use `injectI18n()` HOC. (cc @elastic/kibana-stack-services)
  - It is relevant not just for theming, but any context providers stop working when React is un-mounted.
  - I believe also [React Error Boundaries](https://reactjs.org/docs/error-boundaries.html) and [React Suspense](https://reactjs.org/docs/react-api.html#suspense) stop working when you re-mount React.
- Remounting React is slower performance-wise than just rendering a React component.
- As mentioned above, React will anyways be bundled globally once for everyone; and if somebody wants to render a non-React section, it can be achieved by [this 5-line function](https://github.com/elastic/kibana/pull/43631#discussion_r317835279).

You can still have the "handler pattern", but in the idiomatic React way using render-props:

```tsx
// section/lazy.tsx
export default React.lazy(() => import('.'));

// plugin.tsx
import MySection from './section/lazy';

management.sections.registerLink('my-section', {
  id: 'my-link',
  order: 20,
  renderTitle: () => <>My Link<sup>beta</sup></>
  renderDescription: <p>hello, not used in current UI, but possibly in future</p>
  render: async (context, params) => <MySection context={context} params={params} />;
});
```

- Above the `mount` function is simply renamed to `render` function and returns a `React.ReactNode`. And you can see it uses the standard [React code-splitting pattern with `React.lazy`](https://reactjs.org/docs/code-splitting.html), so the section is loaded dynamically on-demand.
- This way also allows you to use `React.Suspense` and gives ability to the section to control what is displayed while the section is being loaded or to suspend switching for some milliseconds from one section to another while the other section is being loaded; so that navigation from one section to another looks seamless to the user.
- The above pattern is way more powerful, as you can see `renderTitle` and `renderDescription` easily allow you to render anything in those places.

---

About dependence on React. Providing render prop

```tsx
{
  render: async (context, params) => <MySection context={context} params={params} />;
}
```

instead of DOM mounting callback

```tsx
{
  mount: async function mount(context, params) {
    const { renderApp } = await import('./my-section');
    return renderApp(context, params);
  }
}
```

does not make us more dependent on React. In the future, we can still independently remove React from the Management app and independently remove React from the individual sections, regardless if it is `render` rendering function or `mount` handler.

Reply:
> This React re-mounting adds unnecessary code: (1) un-mount React; (2) mount it back; (3) add tests for all of that—which needs to be written and then maintained. 

I could be missing something, but from what I can tell the only different code that is added here is calling `ReactDOM.render`. I'm skeptical this needs to be tested beyond a simple functional test that makes sure you can access the section (which we should already have).

> Then you need to (4) re-attach all your Error Boundaries; (5) re-attach all your React contexts (which I don't think is possible, in a non-trivial cases); (6) all your React Suspense handlers will be probably broken (I'm not aware if it is possible to re-connect them).

I'm not sure I understand what you mean by "re-attach" or why that would break. Yes, each call to the `mount` function would involve re-rendering a React tree. If the UI needed to subscribe to Observables (eg. the UI theme) these would be re-set-up on each call, but that should be a trivial operation. Maybe there's more code here than I realize.

It should be noted, that if a value provided in `context` needs to change it should be provided as an Observable so the React app can subscribe + update when it changes. We should not be unmounting and re-mounting to handle a context value changing.

I'm not sure why React Suspense handlers would break. Some details on this would help.

> Remounting React is slower performance-wise than just rendering a React component.

True, but we only need to do this when navigating to a new section. In that case we're about to render a completely new UI anyways. I don't think there's much performance penalty here but maybe we should experiment and actually test that assumption.

> As mentioned above, React will anyways be bundled globally once for everyone, so a plugin will not need to "bundle React just to render a management section"; and if somebody wants to render a non-React section, it can be achieved by this 5-line function.

Good point, but the answer here is still _it depends_. We plan to provide some out of the box webpack configurations, but it's still entirely likely that 3rd party plugin developers may use a different bundler or config that doesn't know React is already available on the page. In this case, the developer would end up bundling a separate copy of React. Though they could work around this, it'd be nice if we made the right thing easy out of the box.

---

All this to say, I think Vadim's approach is acceptable, it's just that I prefer making Kibana easy to learn over squeaking out a few milliseconds out of the UI.

Too many times in Kibana's history have we invented multiple interfaces & mechanisms for doing very similar things. This has lead to the proliferation of different solutions to the same problem in our source code. It makes learning Kibana development unnecessarily hard.

We need to make Kibana development exceedingly easy. This is as much an internal strategy as it is an external one. Scaling the Kibana & Solution teams at Elastic and growing the plugin community around us is a huge lever in making Kibana an indispensable tool for thousands of users.

Now, is this one thing going to stop someone from picking up Kibana development? Probably not, but the more inconsistencies we add, the worse it becomes. Death by 1000 cuts has to start somewhere.

Reply:
just thought I'd point out that @streamich is out on leave for a bit which is too bad for the sake of this conversation but good for him. 

At risk of speaking without a thorough understanding of all the technical details (which are critically important) - is there a way for us to support framework agnostic _and_ react first ways of doing things? I think we're all in agreement that we expect React to be used the vast majority of the time and our primary reasons for not assuming React usage are -

1) We should avoid assuming 3rd party contributors have a particular knowledge base
2) We wish to avoid more technology coupling than absolutely necessary
3) Tech changes

IMO - these sort of questions become much simpler when we define and agree upon goals.